### PR TITLE
KAFKA-17500: Metadata redirection for NOT_LEADER_OR_FOLLOWER

### PR DESCRIPTION
**Pull request by AndrewJSchofield**
This PR implements the metadata redirection feature of the ShareFetch and ShareAcknowledge responses where an error code of NOT_LEADER_OR_FOLLOWER or FENCED_LEADER_EPOCH along with current leader information in the response is used to optimise handling of leadership changes in the client. This is applying the logic of KIP-951 to share group consumers.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by AndrewJSchofield** [comment]
3 unrelated test failures (features command).

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java [hunk 1]
```diff
@@ -691,7 +712,26 @@ private void handleShareAcknowledgeSuccess(Node fetchTarget,
...
+                                TopicPartition tp = new TopicPartition(metadata.topicNames().get(shareAcknowledgeTopicResponse.topicId()), partitionData.partitionIndex());
+
+                                log.debug("For {}, received error {}, with leaderIdAndEpoch {}", tp, partitionError, partitionData.currentLeader());
+                                if (partitionData.currentLeader().leaderId() != -1 && partitionData.currentLeader().leaderEpoch() != -1) {
+                                    partitionsWithUpdatedLeaderInfo.put(tp, new Metadata.LeaderIdAndEpoch(
+                                        Optional.of(partitionData.currentLeader().leaderId()), Optional.of(partitionData.currentLeader().leaderEpoch())));
+                                }
+
+                                metricsManager.recordFailedAcknowledgements(acknowledgeRequestState.getInFlightAcknowledgementsCount(tip));
+                                acknowledgeRequestState.handleAcknowledgeErrorCode(tip, partitionError);
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: looks these two lines are redundant?. They are getting recorded at line no 739, 740

Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: Thank you. Good catch.

**Review by omkreddy — APPROVED** [review]
@AndrewJSchofield Thanks for the PR. LGTM

**Pull request merged**